src/gtest/test_metrics.cpp

```diff
@@ -107,3 +107,28 @@ TEST(Metrics, EstimateNetHeight) {
     }
     RegtestDeactivateBlossom();
 }
+
+TEST(Metrics, NextUpgrade) {
+    SelectParams(CBaseChainParams::REGTEST);
+    const Consensus::Params& params = Params().GetConsensus();
+
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, 100);
+
+    EXPECT_EQ(SecondsLeftToNextEpoch(params, 1).value(), 14850);
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(params, 1).value(), DurationFormat::REDUCED), "4 hours");
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(params, 1).value(), DurationFormat::FULL), "4 hours, 7 minutes, 30 seconds");
+
+    EXPECT_EQ(SecondsLeftToNextEpoch(params, 90).value(), 1500);
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(params, 90).value(), DurationFormat::REDUCED), "25 minutes");
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(params, 90).value(), DurationFormat::FULL), "25 minutes, 0 seconds");
+
+    EXPECT_EQ(SecondsLeftToNextEpoch(params, 99).value(), 150);
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(params, 99).value(), DurationFormat::REDUCED), "2 minutes");
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(params, 99).value(), DurationFormat::FULL), "2 minutes, 30 seconds");
+
+    auto paramsBlossom = RegtestActivateBlossom(true);
+    EXPECT_EQ(SecondsLeftToNextEpoch(paramsBlossom, 1).value(), 7425);
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(paramsBlossom, 1).value(), DurationFormat::REDUCED), "2 hours");
+    EXPECT_EQ(DisplayDuration(SecondsLeftToNextEpoch(paramsBlossom, 1).value(), DurationFormat::FULL), "2 hours, 3 minutes, 45 seconds");
+    RegtestDeactivateBlossom();
```

Comment:
This looks to have good coverage but I have not checked the heights and expected duration strings.